aptos-move/aptos-gas/src/gas_meter.rs

```diff
@@ -0,0 +1,168 @@
+// Copyright (c) Aptos
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{instr::InstructionGasParameters, transaction::TransactionGasParameters};
+use move_binary_format::{
+    errors::{Location, PartialVMError, PartialVMResult, VMResult},
+    file_format_common::Opcodes,
+};
+use move_core_types::{
+    gas_schedule::{AbstractMemorySize, GasAlgebra, GasCarrier},
+    vm_status::StatusCode,
+};
+use move_vm_types::gas::GasMeter;
+use std::collections::BTreeMap;
+
+pub trait FromOnChainGasSchedule: Sized {
+    fn from_on_chain_gas_schedule(gas_schedule: &BTreeMap<String, u64>) -> Option<Self>;
+}
+
+pub trait ToOnChainGasSchedule {
+    fn to_on_chain_gas_schedule(&self) -> Vec<(String, u64)>;
+}
+
+pub trait InitialGasSchedule: Sized {
+    fn initial() -> Self;
+}
+
+#[derive(Debug, Clone)]
+pub struct NativeGasParameters {
+    pub move_stdlib: move_stdlib::natives::GasParameters,
+    pub aptos_framework: framework::natives::GasParameters,
+}
+
+impl FromOnChainGasSchedule for NativeGasParameters {
+    fn from_on_chain_gas_schedule(gas_schedule: &BTreeMap<String, u64>) -> Option<Self> {
+        Some(Self {
+            move_stdlib: FromOnChainGasSchedule::from_on_chain_gas_schedule(gas_schedule)?,
+            aptos_framework: FromOnChainGasSchedule::from_on_chain_gas_schedule(gas_schedule)?,
```

Comment:
These look exactly the same to me. Why do they need to be 2 different variables?

Reply:
Not totally sure what you mean. These are totally different gas parameters from different packages (move-stdlib, aptos-framework). The move-stdlib one in particular, is defined in the Move repo

Reply:
Oh, maybe you are confused because the two expressions look exactly the same? Notice that `FromOnChainGasSchedule` is a trait, and the two fields have different types that each implements this trait.

Reply:
There's some Rust conversion magic as this calls a specific implementation based upon the destination type.

Reply:
Makes sense this also got me without looking at it in an IDE.